File: 647. Palindromic Substrings.cpp
class Solution {
public:
    int countSubstrings(string s) {
        int sz = s.size();
        bool isPalindrome[sz+1][sz+1];
        int dp[sz+1];
        for( int i = 0 ; i < sz ; i++ ) { 
            for( int j = 0 ; j < sz ; j++ ) isPalindrome[i][j] = false;
        }
        int ans = 0;
        for( int i = 0 ; i < sz ; i++ ) { 
            isPalindrome[i][i] = true;
            ans++;
            for( int x = i, y = i+1 ; y < sz && x >= 0 && s[x] == s[y] ; x--, y++ )  {
                if(isPalindrome[x][y] == false ) { 
                    isPalindrome[x][y] = true;
                    ans++;
                }
            }
            for( int x = i - 1 , y = i + 1 ; y < sz && x >= 0 && s[x] == s[y] ; x--, y++ ) {
                if(isPalindrome[x][y] == false ) { 
                    isPalindrome[x][y] = true;
                    ans++;
                }
            }
        }
        return ans;
        
    }
};

File: python/Lakshay/647. Palindromic Substrings.py
class Solution(object):
    def countSubstrings(self, s):
        """
        :type s: str
        :rtype: int
        
        """
        
   
        ans = 0
        sz = len(s)
        
        for i in range(sz):
            left = i
            right = i + 1
            ans += 1 # each character is a palindrome
            while left >= 0 and right < sz: # calculating even length substring palindrome
                if s[left] == s[right]:
                    left -= 1
                    right += 1
                    ans += 1
                else:
                    break
            
            left = i - 1
            right = i + 1
            while x >= 0 and y < sz: # calcuting odd length substring palindrome
                if s[x] == s[y]:
                    x -= 1
                    y += 1
                    ans += 1
                else:
                    break
        
        return ans
        
